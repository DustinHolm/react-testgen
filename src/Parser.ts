import { Project, PropertySignature, SourceFile, SyntaxKind } from "ts-morph"
import { Attribute, ClassElement, Constructor, Export, ExportType, FunctionElement, ImportBlock, InterfaceElement, MethodElement, VariableElement } from "./types"


class Parser {
    source: SourceFile
    exports: Export[]
    isExportsParsed: boolean

    constructor(path: string) {
        const project = new Project()
        project.addSourceFileAtPath(path)
        const source = project.getSourceFile(path)
        if (source === undefined) {
            throw new Error("SourceFile not found!")
        }

        this.source = source
        this.exports = this.parseExportNames()
        this.isExportsParsed = false
    }

    parseImports(): ImportBlock[] {
        return this.source.getImportDeclarations().map(imDe => {
            let importModule = imDe.getModuleSpecifier().getText()
            if (importModule.startsWith("\"") && importModule.endsWith("\"")
                || importModule.startsWith("\'") && importModule.endsWith("\'")) {
                importModule = importModule.slice(1, -1)
            }
            const importModuleIsInternal = importModule.startsWith(".")

            const namedImports = imDe.getNamedImports().map(imSp =>
                imSp.getName())
            const defaultImport = imDe.getDefaultImport()?.getText()

            const imports: Export[] = []

            namedImports.forEach(naIm =>
                imports.push({
                    name: naIm,
                    isDefault: false,
                })
            )

            if (defaultImport !== undefined) {
                imports.push({
                    name: defaultImport,
                    isDefault: true,
                })
            }

            return {
                sourceFile: importModule,
                isInternal: importModuleIsInternal,
                imports: imports
            }
        })
    }

    parseExports(): Export[] {
        if (!this.isExportsParsed) {
            this.parseInterfaces()
            this.parseClasses()
            this.parseFunctions()
            this.parseVariables()
            this.isExportsParsed = true
        }

        return this.exports
    }

    parseInterfaces() {
        this.source.getInterfaces().forEach(inDe => {
            const existingExport = this.exports.find(e => e.name === inDe.getName())

            if (existingExport !== undefined) {
                const attributes: Attribute[] = inDe.getMembers().map(tyElTy => ({
                    name: (tyElTy as PropertySignature).getName(),
                    type: tyElTy.getType().getText()
                }))

                existingExport.element = {
                    returnsJSX: false,
                    type: ExportType.Interface,
                    attributes: attributes
                } as InterfaceElement
            }
        })
    }

    parseClasses() {
        this.source.getClasses().forEach(clDe => {
            const existingExport = this.exports.find(e => e.name === clDe.getName())

            if (existingExport !== undefined) {
                const constructors: Constructor[] = clDe.getConstructors().map(coDe => {
                    const attributes: Attribute[] = coDe.getParameters().map(p => ({
                        name: p.getName(),
                        type: p.getType().getText()
                    }))

                    return {
                        parameters: attributes
                    }
                })


                const methods = clDe.getMethods().map(meDe => {
                    const name = meDe.getName()
                    const returnType = meDe.getReturnType().getText()

                    const parameters: Attribute[] = meDe.getParameters().map(paDe => ({
                        name: paDe.getName(),
                        type: paDe.getType().getText()
                    }))

                    return {
                        returnsJSX: returnType.includes("JSX"),
                        type: ExportType.Function,
                        returnType: returnType,
                        parameters: parameters,
                        name: name
                    } as MethodElement
                })

                existingExport.element = {
                    returnsJSX: methods.some(m => m.returnsJSX),
                    type: ExportType.Class,
                    methods: methods,
                    constructors: constructors
                } as ClassElement
            }
        })
    }

    parseFunctions() {
        this.source.getFunctions().forEach(fuDe => {
            const existingExport = this.exports.find(e => e.name === fuDe.getName())

            if (existingExport !== undefined) {
                const returnType = fuDe.getReturnType().getText()

                const parameters: Attribute[] = fuDe.getParameters().map(paDe => ({
                    name: paDe.getName(),
                    type: paDe.getType().getText()
                }))

                existingExport.element = {
                    returnsJSX: returnType.includes("JSX"),
                    type: ExportType.Function,
                    returnType: returnType,
                    parameters: parameters
                } as FunctionElement
            }
        })
    }

    parseVariables() {
        this.source.getVariableDeclarations().forEach(vaDe => {
            const existingExport = this.exports.find(e => e.name === vaDe.getName())

            if (existingExport !== undefined) {
                const aliasType = vaDe.getChildrenOfKind(SyntaxKind.TypeReference)
                    .map(typeRef => typeRef.getText())
                    .shift()


                const variableType = aliasType ?? vaDe.getType().getText()
                let variableValue = undefined
                let parameters: Attribute[] | undefined = undefined
                let returnType: string | undefined = undefined
                const arrowFunction = vaDe.getInitializerIfKind(SyntaxKind.ArrowFunction)
                const otherFunction = vaDe.getInitializerIfKind(SyntaxKind.FunctionExpression)

                if (arrowFunction) {
                    returnType = arrowFunction.getReturnType().getText()

                    parameters = arrowFunction.getParameters().map(p => ({
                        name: p.getName(),
                        type: p.getType().getText()
                    }))
                } else if (otherFunction) {
                    returnType = otherFunction.getReturnType().getText()

                    parameters = otherFunction.getParameters().map(p => ({
                        name: p.getName(),
                        type: p.getType().getText()
                    }))
                } else {
                    variableValue = vaDe.getInitializer()?.getText()
                }

                if (parameters !== undefined && returnType !== undefined) {
                    existingExport.element = {
                        returnsJSX: returnType.includes("JSX"),
                        type: ExportType.Function,
                        returnType: returnType,
                        parameters: parameters
                    } as FunctionElement
                } else {
                    existingExport.element = {
                        returnsJSX: variableType.includes("JSX"),
                        type: ExportType.Variable,
                        varType: variableType,
                        varValue: variableValue ?? "undefined"
                    } as VariableElement
                }
            }
        })
    }

    private parseExportNames(): Export[] {
        const defaultExports = this.source.getExportAssignments()
        const exportSymbols = this.source.getExportSymbols()
        const exportNames: Export[] = []

        defaultExports.forEach(exAs => {
            exportNames.push({
                name: exAs.getExpression().getText(),
                isDefault: true
            })
        })

        exportSymbols.forEach(exSy => {
            if (exSy.getName() !== "default") {
                exportNames.push({
                    name: exSy.getName(),
                    isDefault: false
                })
            }
        })

        return exportNames
    }
}

export default Parser