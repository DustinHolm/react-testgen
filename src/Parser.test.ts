import Parser from "./Parser"

describe("Parser Tests", () => {
    test("Nonexisting file throws error", () => {
        const parser = new Parser("./__test_resources__/ReactClass.tsx")
        expect(parser.parse).toThrow()
    })

    test("ReactClass matches snapshot", () => {
        const parser = new Parser("./__test_resources__/ReactClass.tsx")
        const parserResult = parser.parse()
        expect(parserResult).toMatchSnapshot()
    })

    test("ReactFC matches snapshot", () => {
        const parser = new Parser("./__test_resources__/ReactFC.tsx")
        const parserResult = parser.parse()
        expect(parserResult).toMatchSnapshot()
    })

    test("ReactFunction matches snapshot", () => {
        const parser = new Parser("./__test_resources__/ReactFunction.tsx")
        const parserResult = parser.parse()
        expect(parserResult).toMatchSnapshot()
    })

    test("ReactClassDependency matches snapshot", () => {
        const parser = new Parser("./__test_resources__/ReactClassDependency.tsx")
        const parserResult = parser.parse()
        expect(parserResult).toMatchSnapshot()
    })

    test("ReactFCDependency matches snapshot", () => {
        const parser = new Parser("./__test_resources__/ReactFCDependency.tsx")
        const parserResult = parser.parse()
        expect(parserResult).toMatchSnapshot()
    })
})