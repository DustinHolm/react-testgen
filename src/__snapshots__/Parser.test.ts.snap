// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser Tests ReactClass matches snapshot 1`] = `
Object {
  "exports": Array [
    Object {
      "element": Object {
        "constructors": Array [
          Object {
            "parameters": Array [
              Object {
                "name": "props",
                "type": "import(\\"/media/dustin/Data/Code/react-testgen/__test_resources__/ReactClass\\").ReactClassProps",
              },
            ],
          },
        ],
        "methods": Array [
          Object {
            "name": "increase",
            "parameters": Array [],
            "returnType": "void",
            "returnsJSX": false,
            "type": 1,
          },
          Object {
            "name": "render",
            "parameters": Array [],
            "returnType": "JSX.Element",
            "returnsJSX": true,
            "type": 1,
          },
        ],
        "returnsJSX": true,
        "type": 0,
      },
      "isDefault": true,
      "name": "ReactClass",
    },
    Object {
      "element": Object {
        "returnsJSX": false,
        "type": 3,
        "varType": "number",
        "varValue": "2",
      },
      "isDefault": false,
      "name": "initialStateCount",
    },
    Object {
      "element": Object {
        "attributes": Array [
          Object {
            "name": "content",
            "type": "string",
          },
        ],
        "returnsJSX": false,
        "type": 2,
      },
      "isDefault": false,
      "name": "ReactClassProps",
    },
  ],
  "imports": Array [
    Object {
      "imports": Array [
        Object {
          "isDefault": false,
          "name": "Component",
        },
      ],
      "isInternal": false,
      "sourceFile": "react",
    },
    Object {
      "imports": Array [
        Object {
          "element": Object {
            "constructors": Array [
              Object {
                "parameters": Array [
                  Object {
                    "name": "props",
                    "type": "ReactClassDependencyProps",
                  },
                ],
              },
            ],
            "methods": Array [
              Object {
                "name": "render",
                "parameters": Array [],
                "returnType": "JSX.Element",
                "returnsJSX": true,
                "type": 1,
              },
            ],
            "returnsJSX": true,
            "type": 0,
          },
          "isDefault": true,
          "name": "ReactClassDependency",
        },
      ],
      "isInternal": true,
      "sourceFile": "./ReactClassDependency",
    },
    Object {
      "imports": Array [
        Object {
          "element": Object {
            "parameters": Array [
              Object {
                "name": "value",
                "type": "number",
              },
            ],
            "returnType": "number",
            "returnsJSX": false,
            "type": 1,
          },
          "isDefault": false,
          "name": "utilitySquareFunction",
        },
      ],
      "isInternal": true,
      "sourceFile": "./utilityFunctions",
    },
  ],
}
`;

exports[`Parser Tests ReactClassDependency matches snapshot 1`] = `
Object {
  "exports": Array [
    Object {
      "element": Object {
        "constructors": Array [
          Object {
            "parameters": Array [
              Object {
                "name": "props",
                "type": "ReactClassDependencyProps",
              },
            ],
          },
        ],
        "methods": Array [
          Object {
            "name": "render",
            "parameters": Array [],
            "returnType": "JSX.Element",
            "returnsJSX": true,
            "type": 1,
          },
        ],
        "returnsJSX": true,
        "type": 0,
      },
      "isDefault": true,
      "name": "ReactClassDependency",
    },
  ],
  "imports": Array [
    Object {
      "imports": Array [
        Object {
          "isDefault": false,
          "name": "Component",
        },
      ],
      "isInternal": false,
      "sourceFile": "react",
    },
  ],
}
`;

exports[`Parser Tests ReactFC matches snapshot 1`] = `
Object {
  "exports": Array [
    Object {
      "element": Object {
        "parameters": Array [
          Object {
            "name": "props",
            "type": "React.PropsWithChildren<import(\\"/media/dustin/Data/Code/react-testgen/__test_resources__/ReactFC\\").ReactFCProps>",
          },
        ],
        "returnType": "JSX.Element",
        "returnsJSX": true,
        "type": 1,
      },
      "isDefault": true,
      "name": "ReactFC",
    },
    Object {
      "element": Object {
        "returnsJSX": false,
        "type": 3,
        "varType": "number",
        "varValue": "2",
      },
      "isDefault": false,
      "name": "initialStateCount",
    },
    Object {
      "element": Object {
        "attributes": Array [
          Object {
            "name": "content",
            "type": "string",
          },
        ],
        "returnsJSX": false,
        "type": 2,
      },
      "isDefault": false,
      "name": "ReactFCProps",
    },
  ],
  "imports": Array [
    Object {
      "imports": Array [
        Object {
          "isDefault": false,
          "name": "FC",
        },
        Object {
          "isDefault": false,
          "name": "useState",
        },
      ],
      "isInternal": false,
      "sourceFile": "react",
    },
    Object {
      "imports": Array [
        Object {
          "element": Object {
            "parameters": Array [
              Object {
                "name": "props",
                "type": "React.PropsWithChildren<ReactFCDependencyProps>",
              },
            ],
            "returnType": "JSX.Element",
            "returnsJSX": true,
            "type": 1,
          },
          "isDefault": true,
          "name": "ReactFCDependency",
        },
      ],
      "isInternal": true,
      "sourceFile": "./ReactFCDependency",
    },
    Object {
      "imports": Array [
        Object {
          "element": Object {
            "parameters": Array [
              Object {
                "name": "value",
                "type": "number",
              },
            ],
            "returnType": "number",
            "returnsJSX": false,
            "type": 1,
          },
          "isDefault": false,
          "name": "utilitySquareFunction",
        },
      ],
      "isInternal": true,
      "sourceFile": "./utilityFunctions",
    },
  ],
}
`;

exports[`Parser Tests ReactFCDependency matches snapshot 1`] = `
Object {
  "exports": Array [
    Object {
      "element": Object {
        "parameters": Array [
          Object {
            "name": "props",
            "type": "React.PropsWithChildren<ReactFCDependencyProps>",
          },
        ],
        "returnType": "JSX.Element",
        "returnsJSX": true,
        "type": 1,
      },
      "isDefault": true,
      "name": "ReactFCDependency",
    },
  ],
  "imports": Array [
    Object {
      "imports": Array [
        Object {
          "isDefault": false,
          "name": "FC",
        },
      ],
      "isInternal": false,
      "sourceFile": "react",
    },
  ],
}
`;

exports[`Parser Tests ReactFunction matches snapshot 1`] = `
Object {
  "exports": Array [
    Object {
      "element": Object {
        "parameters": Array [
          Object {
            "name": "props",
            "type": "import(\\"/media/dustin/Data/Code/react-testgen/__test_resources__/ReactFunction\\").ReactFunctionProps",
          },
        ],
        "returnType": "JSX.Element",
        "returnsJSX": true,
        "type": 1,
      },
      "isDefault": true,
      "name": "ReactFunction",
    },
    Object {
      "element": Object {
        "returnsJSX": false,
        "type": 3,
        "varType": "number",
        "varValue": "2",
      },
      "isDefault": false,
      "name": "initialStateCount",
    },
    Object {
      "element": Object {
        "attributes": Array [
          Object {
            "name": "content",
            "type": "string",
          },
        ],
        "returnsJSX": false,
        "type": 2,
      },
      "isDefault": false,
      "name": "ReactFunctionProps",
    },
  ],
  "imports": Array [
    Object {
      "imports": Array [
        Object {
          "isDefault": false,
          "name": "useState",
        },
      ],
      "isInternal": false,
      "sourceFile": "react",
    },
    Object {
      "imports": Array [
        Object {
          "element": Object {
            "parameters": Array [
              Object {
                "name": "props",
                "type": "React.PropsWithChildren<ReactFCDependencyProps>",
              },
            ],
            "returnType": "JSX.Element",
            "returnsJSX": true,
            "type": 1,
          },
          "isDefault": true,
          "name": "ReactFCDependency",
        },
      ],
      "isInternal": true,
      "sourceFile": "./ReactFCDependency",
    },
    Object {
      "imports": Array [
        Object {
          "element": Object {
            "parameters": Array [
              Object {
                "name": "value",
                "type": "number",
              },
            ],
            "returnType": "number",
            "returnsJSX": false,
            "type": 1,
          },
          "isDefault": false,
          "name": "utilitySquareFunction",
        },
      ],
      "isInternal": true,
      "sourceFile": "./utilityFunctions",
    },
  ],
}
`;
